src/stub_read.s

```diff
@@ -1,47 +1,60 @@
-              .extern _Zp
+            .extern _Zp
 
-CHKIN:	      .equlab 0xffc6
-CHRIN:	      .equlab 0xffcf
-READST:	      .equlab 0xffb7
-CLRCHN:       .equlab 0xffcc
+CHKIN:	    .equlab 0xffc6
+CHRIN:	    .equlab 0xffcf
+READST:	    .equlab 0xffb7
+CLRCHN:     .equlab 0xffcc
 
-              .section code
-              .public _Stub_read
-_Stub_read:   ldx     zp:_Zp+0
-	      beq     start$	    ; stdin
-	      cpx     #3	    ; test for stdout/stderr
-	      bcc     eof$
-	      jsr     CHKIN	    ; ensure channel is prepared to read
-start$:	      lda     #0	    ; set file counter = 0
-              sta     zp:_Zp+0
-              sta     zp:_Zp+1
-              ldx     zp:_Zp+5      ; 256 or more?
-              beq     20$           ; no
-              ldy     #0
-10$:          jsr     CHRIN
-              sta     (_Zp+2),y
-              iny
-              bne     10$
-              inc     zp:_Zp+3      ; step 256
-              inc     zp:_Zp+1
-              dec     zp:_Zp+5      ; yes, decrement 256 counter
-              bne     10$
+            .section code
+            .public _Stub_read
+_Stub_read: ldx     zp:_Zp+0
+            beq     start$	    ; stdin
+            cpx     #3	    ; test for stdout/stderr
+            bcc     eof$
+            jsr     CHKIN	    ; ensure channel is prepared to read
```

Comment:
Looking at the documentation, I think it should call `READST` after `CHKIN` and 64 means end of file (EOI). Zero is OK, and any other number is an error condition.

Something like:
```
        jsr    CHKIN
        jsr    READST
        tax
        bne  exit
; do normal stuff
.
.
exit:   cmp   #64
        beq   eof$
; some I/O error, set errno, return -1
           ...
```

Reply:
It's slightly more complicated than that. The EOI indication is transmitted by the disk drive immediately BEFORE it sends the last byte. (It's sent by delaying the CLK line being pulled true by 200us). So, READST returns 64 when there is still one byte to hand up the stack.

If the program attempt further reads, it returns 66.

Put it another way, assume that there is a one byte file. And then the program does::

uint8_t buffer[100];
size_t bytes_read;
do {
   bytes_read = _Stub_read(testfile_fd, buffer, 100);
} while (bytes_read > 0);

Which is the canonical loop in C to read the entire file.

The first time around, bytes_read needs to be 1, so that the program processes that one byte. So, even though I know that we're at the end of file, I don't have any C way to TELL anyone that in this pass. Then it has to return 0 or -1 on the next call. But what is the commodore doing at that point?

First, _Stub_read calls CHKIN to redirect input from that file. Then it calls CHRIN/BASIN to get the character. The read pointer is now past the end of the file. Thus, READST now returns 66, which is both EOI and "timeout error". But this is the point at which we need to pass up the actual EOF indication. If I set the return value to -1, then I'd have to set errno to something that makes sense to the rest of stdio. I don't see a code in errno.h for that.

What I do see is that feof want's a file flag set:
return (stream->fs_flags & __FS_FLAG_EOF) != 0;

That's set by __lib_read on a short read:
shortread:
#endif

      /* Set or clear the EOF indicator.  If we get here because of a
       * short read and the total number of bytes read is zero, then
       * we must be at the end-of-file.
       */

      if (remaining == 0)
        {
          stream->fs_flags &= ~__FS_FLAG_EOF;
        }
      else
        {
          stream->fs_flags |= __FS_FLAG_EOF;
        }

      /* lib_give_semaphore(stream); */
      return count - remaining;
    }


But can we get there this way?
The code that runs when I return -1 would be, I think:
bytes_read = _Stub_read(stream->fs_fd, dest, remaining);
if (bytes_read < 0)
  {
    if (count - remaining > 0)
      {
        break;
      }
    else
      {
        /* An error occurred on the read.  The error code is
         * in the 'errno' variable.
         */

        __set_errno(bytes_read);
        goto errout_with_errno;
      }
  }

Now, remember the implication of the above.... _Stub_read can only return an EOF indication on the read attempt AFTER the last valid byte is delivered. If we change our test program to read:
uint8_t buffer[1];
size_t bytes_read;
do {
   bytes_read = fread(buffer, 1, 1, testfile_fp);
} while (bytes_read > 0);

That means that fread will get one call that must return one byte, and then a call with no data to deliver. Count will be 1, and remaining will also be 1. So, we go down the path where we set the errno system variable, and then go to errout_with_errno.

That does:
errout_with_errno:
  stream->fs_flags |= __FS_FLAG_ERROR;
  /* lib_give_semaphore(stream); */
  return ERROR;
  
Which sets the wrong fs flag. feof would not return true in this case.

The only way to get lib_fread to set __FS_FLAG_EOF is to return 0 if READST returns 66 on the first byte read from CHRIN. While I agree that calling fread on one byte buffers would be a poor programming practice, this issue will occur on any time where the file length is some multiple of the read buffer size.

Reply:
Ouch, that was a bit complicated.

I can see that `__lib_fread()` sets `errno` basd on the number of bytes read last time (the original code negates it even), which to me makes no sense at all. Well it gets a number and as `read` can return short, it seems it has no well defined meaning. The rest of the stub calls expects that if there is a problem, return `-1` and 
set `errno` to some error code. But this one is different.

I do not bother so much about the error code put in `errno`, a suitable one from the table is good, else pick something that seems to make some sense.

I will change the `__lib_fread()` on my side not to set `errno`, and rely on the stub to do it (if it wants to communicate something). As it is not it writes it over.

If you are happy with the Commodore protocol I am going to trust you. I notice that there are some tabs in the file, I will fix that later. I will also add an error code if someone tries to read from `stdin`.